### PR TITLE
[Feat] Add cargo package search

### PR DESCRIPTION
**Pull request by dkuku**
## What does this PR do?

Add cargo package search
this depends on changes from #3464 
image

### CODE REVIEW

File: searx/engines/cargo.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    docs_key = gettext("Documentation")
+    homepage_key = gettext("Project homepage")
+    source_code_key = gettext("Source code")
+    links = {}
+    if package.get("documentation"):
+        links[docs_key] = package["documentation"]
+    if package.get("homepage"):
+        links[homepage_key] = package["homepage"]
+    if package.get("repository"):
+        links[source_code_key] = package["repository"]
```
Comment (Bnyro) [inline searx/engines/cargo.py]:
```suggestion
    if "documentation" in package:
        links[docs_key] = package["documentation"]
    if "homepage" in package:
        links[homepage_key] = package["homepage"]
    if "repository" in package:
        links[source_code_key] = package["repository"]
```

File: searx/engines/cargo.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    if links:
+        return links
```
Comment (Bnyro) [inline searx/engines/cargo.py]: This should probably return `links`, no matter if it's empty or not (not sure how Python handles functions without return values)

File: searx/engines/cargo.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+"""hex.pm"""
+
+from urllib.parse import urlencode
+from dateutil import parser
+from flask_babel import gettext
+
+
+about = {
+    # pylint: disable=line-too-long
```
Comment (Bnyro) [inline searx/engines/cargo.py]: nit: not needed here because the api docs line is short enough

File: searx/engines/cargo.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+categories = ["it", "packages"]
+
+
+# engine dependent config
+paging = True
+search_url = "https://crates.io/api/v1/crates"
+
+
+def request(query: str, params):
+    args = urlencode({"page": params["pageno"], "q": query, "per_page": 20})
```
Comment (Bnyro) [inline searx/engines/cargo.py]: per_page could be made configurable in `settings.yml` by making it a global variable called `page_size`, but that's not mandatory

**Review by Bnyro — APPROVED** [review]

File: searx/engines/crates.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+about = {
+    "website": "https://crates.io/",
+    "wikidata_id": None,
+    "official_api_documentation": "https://crates.io/data-access",
+    "use_official_api": True,
+    "require_api_key": False,
+    "results": "JSON",
+}
+
+categories = ["it", "packages", "cargo"]
```
Comment (return42) [inline searx/engines/crates.py:18]: You can further use `!cargo` .. when we implement other cargo reposetories we will add this engines also to group `cargo` .. so with `!cargo` we are able to search through all cargo managers in once.

File: searx/engines/crates.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+
+# engine dependent config
+paging = True
+page_size = 10
```
Comment (return42) [inline searx/engines/crates.py:23]: As @Bnyro mentioned .. the page_size is configurable and I reduced from 20 to 10 .. 10 results per page is the default what we have in nearly all engines.

**Review by return42 — APPROVED** [review]
I renamed the module from `cargo.py` to `crates.io` .. "cargo" is the package manager while "crates.io" is just one package repository / may we want to implement for example https://lib.rs/ at one day.

Due to the rebase on

- https://github.com/searxng/searxng/pull/3464

and the rename to `crates.py` this PR has changed significantly .. can you please review again / thanks!

File: searx/settings.yml [hunk 7]
```diff
@@ -924,6 +924,12 @@ engines:
     shortcut: hex
     disabled: true
 
+  - name: crates.io
+    engine: crates
+    shortcut: crates
+    disabled: true
+    timeout: 6.0
```
Comment (return42) [inline searx/settings.yml:931]: Forgot to say: I had timeout issues on search terms like `!cargo test` or `!cargo lib` .. so I increased timeout from the default 3sec to 6sec .. not sure this is a good value.

**Comment by dkuku** [comment]
Can this be merged? I don't have permissions.

**Comment by return42** [comment]
> Can this be merged? I don't have permissions.

sure .. I was just waiting for the OK from you for my modifications :-)

**Pull request merged**